### PR TITLE
fix(provider): Gemma 4 VLM continuous-batched decode — no more repetition

### PR DESCRIPTION
**Pull request by Gajesh2007**
## Summary

Fixes the Gemma 4 **token-repetition loops** users hit on the chat console (e.g. `"...the era of the era of..."`, `"One of of of of"`, `"you've you've you've..."`).

`gemma-4-26b` declares `vision_config`, so providers load it via `VLMModelFactory` and serve **text** through the VLM model's continuous-batched path. That path had two bugs that compound on 4-bit builds:

1. **Scalar RoPE offset** — used `cache.offset` (max across rows), mis-positioning shorter rows in a mixed-length batch.
2. **Explicit-mask fused-kernel divergence** ([ml-explore/mlx#3384](https://github.com/ml-explore/mlx/issues/3384)) — `MLXFast.scaledDotProductAttention` numerically diverges with an explicit array mask on 4-bit Gemma 4, flipping top-1 logits.

The earlier v0.6.9 NaN fix targeted `MLXLLM/Gemma4Text.swift` — a class providers don't load for vision checkpoints — which is why it didn't help.

The actual code fix is in the submodule: **Layr-Labs/mlx-swift-lm#41**. This PR bumps the submodule pointer to that fix and adds regression coverage.

## Changes

- **`libs/mlx-swift-lm`**: `404afee → 429d83c` (per-row RoPE offset + NaN-guarded manual masked attention in `MLXVLM/Models/Gemma4.swift`; `.none` decode mask in `BatchKVCache`). See Layr-Labs/mlx-swift-lm#41.
- **`ContinuousBatchingLiveTests.gemma4VLMMixedLengthCoherent`** — loads `gemma-4-26B-A4B-it-qat-4bit` via `VLMModelFactory` (the production path; the existing `runDiffTest` used `LLMModelFactory`, which is why it never caught this) and asserts a mixed-length `B=2` greedy batch is coherent. Gated by `DARKBLOOM_LIVE_MLX_TESTS=1` + `DARKBLOOM_LIVE_MLX_GEMMA=1`.
- **`BatchKVCacheTests`** — adds a case asserting `.none` mask for unpadded `n==1` decode and `.array` when any row is left-padded.

## Verification

On `gemma-4-26B-A4B-it-qat-4bit` via `VLMModelFactory`, mixed-length `B=2` greedy:

- **Before:** `"...the era of the era of the era of the era of..."`
- **After:** coherent (full "Shortcut Learning" answer + a coherent renewable-energy essay).

A/B verified by reverting the submodule fix (repetition returns → test fails) and re-applying (coherent → test passes). All `BatchKVCache` unit tests pass.

## Merge order

Merge the submodule PR (Layr-Labs/mlx-swift-lm#41) first, then re-point `libs/mlx-swift-lm` here to the merged SHA if it differs (squash-merge changes the hash). After merge, **bump the provider version** so the fleet picks up the fix (`LatestProviderVersion` gate).

---
<a href="https://app.blacksmith.sh/Layr-Labs/codesmith/d-inference/pr/369"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1784225487&installation_id=133247490&pr_number=369&repository=Layr-Labs%2Fd-inference&return_to=https%3A%2F%2Fgithub.com%2FLayr-Labs%2Fd-inference%2Fpull%2F369&signature=763a54e71aca1e7df4b9ac2d7d4b46cde277752bc472b676eab1f9675d0ae69a"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 1]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+        let maxTokens = 110
+
+        let batched = try await runBatchedEngine(
+            container: container, modelID: modelID, prompts: encoded, maxTokens: maxTokens)
+        let single = await singleStreamGreedy(
+            container: container, prompts: encoded, maxTokens: maxTokens)
+
+        // The batched engine honors EOS, so coherent rows terminate cleanly.
+        // This is the path under test (per-row offset + manual masked attention);
+        // it must not degenerate into repetition.
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:130]: 🟡 **[MEDIUM]** ⚡ Redundant tokenizer.applyChatTemplate calls in map - could pre-compute template once

💡 _Suggestion:_ Extract chat template application outside the map or cache the template format to avoid repeated template processing for each prompt

📊 _Score: 3×4 = 12 · Category: Repeated work_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 2]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+    /// Detects degenerate generation: an n-gram (n = 1...4) repeated
+    /// consecutively many times — the signature of decode loops like
+    /// "of of of", "you've you've", or "the era of the era of the era of".
+    /// Coherent prose never trips this; the batching bug produces exactly
+    /// these cycles.
+    static func hasDegenerateRepetition(_ tokens: [Int]) -> Bool {
+        for n in 1 ... 4 {
+            let minReps = n == 1 ? 8 : (n == 2 ? 6 : 4)
+            var i = 0
+            while i + n * minReps <= tokens.count {
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:162]: 🔵 **[INFO]** ⚡ O(n²) nested loop in hasDegenerateRepetition for pattern matching

💡 _Suggestion:_ Consider using a rolling hash or KMP algorithm for pattern detection, especially for longer token sequences

📊 _Score: 2×3 = 6 · Category: Inefficient data structures_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 3]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+                let gram = Array(tokens[i ..< i + n])
+                var reps = 1
+                var j = i + n
+                while j + n <= tokens.count && Array(tokens[j ..< j + n]) == gram {
+                    reps += 1
+                    j += n
+                }
+                if reps >= minReps { return true }
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:170]: 🔵 **[INFO]** ⚡ Array slice allocations in tight loop without pre-allocation

💡 _Suggestion:_ Pre-allocate gram array outside the loop and reuse it, or use array indices instead of creating new slices

📊 _Score: 2×3 = 6 · Category: Unbounded allocations_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 4]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+        let container = try await VLMModelFactory.shared.loadContainer(
+            from: modelDir, using: LocalTokenizerLoader())
+
+        // Deliberately different lengths so the shorter row carries left-padding
+        // (the case the scalar-offset bug mis-positions). Row 0 is the shorter,
+        // higher-entropy prompt — open-ended continuations have close argmax
+        // calls, so wrong RoPE positions / mask-kernel drift flip the top token
+        // and trap it in a loop (the "One of of of of" failure mode).
+        let prompts = [
+            "Tell me something interesting about machine learning.",
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:110]: 🟡 **[MEDIUM]** 🧩 Complex test setup duplicates production loading logic unnecessarily

💡 _Suggestion:_ Extract a shared helper for VLM container loading or use dependency injection to avoid duplicating production code paths in tests

📊 _Score: 3×4 = 12 · Category: over-abstraction_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 5]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+                while j + n <= tokens.count && Array(tokens[j ..< j + n]) == gram {
+                    reps += 1
+                    j += n
+                }
+                if reps >= minReps { return true }
+                i += 1
+            }
+        }
+        return false
+    }
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:175]: 🔵 **[INFO]** 🧩 Degenerate repetition detection could be extracted to shared utility

💡 _Suggestion:_ Move hasDegenerateRepetition to a shared test utility module since this pattern detection logic could be useful across multiple test files

📊 _Score: 2×3 = 6 · Category: duplicate logic_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 6]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+            shortMatch >= 3,
+            Comment(rawValue:
+                "short row diverges immediately (batched=\(batchedHead) single=\(singleHead))"))
+    }
+
+    /// Detects degenerate generation: an n-gram (n = 1...4) repeated
+    /// consecutively many times — the signature of decode loops like
+    /// "of of of", "you've you've", or "the era of the era of the era of".
+    /// Coherent prose never trips this; the batching bug produces exactly
+    /// these cycles.
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:157]: 🔵 **[INFO]** 🧩 Test has multiple environment variable dependencies and hardcoded values

💡 _Suggestion:_ Consolidate environment checks into a single test enablement flag and extract hardcoded values (maxTokens=110, minReps thresholds) to constants

📊 _Score: 2×3 = 6 · Category: over-configuration_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — ✅ No issues found

### Performance — 3 finding(s) (1 blocking)
- 🟡 **[MEDIUM]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:126-130` — Redundant tokenizer.applyChatTemplate calls in map - could pre-compute template once
  - _Suggestion:_ Extract chat template application outside the map or cache the template format to avoid repeated template processing for each prompt
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:158-162` — O(n²) nested loop in hasDegenerateRepetition for pattern matching
  - _Suggestion:_ Consider using a rolling hash or KMP algorithm for pattern detection, especially for longer token sequences
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:164-170` — Array slice allocations in tight loop without pre-allocation
  - _Suggestion:_ Pre-allocate gram array outside the loop and reuse it, or use array indices instead of creating new slices

### Type_diligence — ✅ No issues found

### Additive_complexity — 3 finding(s) (1 blocking)
- 🟡 **[MEDIUM]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:89-110` — Complex test setup duplicates production loading logic unnecessarily
  - _Suggestion:_ Extract a shared helper for VLM container loading or use dependency injection to avoid duplicating production code paths in tests
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:158-175` — Degenerate repetition detection could be extracted to shared utility
  - _Suggestion:_ Move hasDegenerateRepetition to a shared test utility module since this pattern detection logic could be useful across multiple test files
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:111-157` — Test has multiple environment variable dependencies and hardcoded values
  - _Suggestion:_ Consolidate environment checks into a single test enablement flag and extract hardcoded values (maxTokens=110, minReps thresholds) to constants

_6 finding(s) total, 2 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift (hunk 1 above)
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:130]: 🔵 **[INFO]** ⚡ Redundant tokenizer.applyChatTemplate calls in map - could pre-compute template

💡 _Suggestion:_ Consider caching the chat template result or restructuring to avoid repeated template application if the template format is identical across prompts

📊 _Score: 2×3 = 6 · Category: repeated_work_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 7]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+                var j = i + n
+                while j + n <= tokens.count && Array(tokens[j ..< j + n]) == gram {
+                    reps += 1
+                    j += n
+                }
+                if reps >= minReps { return true }
+                i += 1
+            }
+        }
+        return false
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:174]: 🟡 **[MEDIUM]** ⚡ Nested loops in hasDegenerateRepetition create O(n²×k) complexity for token analysis

💡 _Suggestion:_ Use a sliding window approach or KMP-style pattern matching to reduce complexity from O(n²×k) to O(n×k) where n is token count and k is max n-gram size

📊 _Score: 3×4 = 12 · Category: inefficient_algorithm_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift [hunk 8]
```diff
@@ -69,6 +70,110 @@ struct ContinuousBatchingLiveTests {
...
+        // past EOS, so only the pre-EOS prefix is meaningful). With per-row
+        // offsets the two agree on the opening tokens.
+        let eos = 106
+        let singleHead = Array(single[0].prefix(while: { $0 != eos }))
+        let batchedHead = Array(batched[0].prefix(while: { $0 != eos }))
+        let shortMatch = zip(batchedHead, singleHead).prefix(while: ==).count
+        #expect(
+            shortMatch >= 3,
+            Comment(rawValue:
+                "short row diverges immediately (batched=\(batchedHead) single=\(singleHead))"))
```
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:150]: 🔵 **[INFO]** 🧩 Complex test setup with multiple environment variables and fallback logic

💡 _Suggestion:_ Consider simplifying the test configuration - the nested environment variable checks and fallback model ID logic could be streamlined into a helper function or consolidated into fewer configuration points

📊 _Score: 2×3 = 6 · Category: over-configuration_

File: provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift (hunk 7 above)
Comment (ethenotethan) [inline provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:174]: 🔵 **[INFO]** 🧩 Hardcoded test data that could be parameterized or shared

💡 _Suggestion:_ The prompt construction and encoding logic is specific but could potentially be extracted to a test helper if similar patterns emerge in other tests

📊 _Score: 2×2 = 4 · Category: duplicate logic_

**Review by ethenotethan — CHANGES_REQUESTED** [review]
## Automated Code Review — Layr-Labs/d-inference#

**Verdict:** REQUEST_CHANGES

### Security — ✅ No issues found

### Performance — 2 finding(s) (1 blocking)
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:126-130` — Redundant tokenizer.applyChatTemplate calls in map - could pre-compute template
  - _Suggestion:_ Consider caching the chat template result or restructuring to avoid repeated template application if the template format is identical across prompts
- 🟡 **[MEDIUM]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:158-174` — Nested loops in hasDegenerateRepetition create O(n²×k) complexity for token analysis
  - _Suggestion:_ Use a sliding window approach or KMP-style pattern matching to reduce complexity from O(n²×k) to O(n×k) where n is token count and k is max n-gram size

### Type_diligence — ✅ No issues found

### Additive_complexity — 3 finding(s)
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:126-150` — Complex test setup with multiple environment variables and fallback logic
  - _Suggestion:_ Consider simplifying the test configuration - the nested environment variable checks and fallback model ID logic could be streamlined into a helper function or consolidated into fewer configuration points
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:158-174` — Hardcoded test data that could be parameterized or shared
  - _Suggestion:_ The prompt construction and encoding logic is specific but could potentially be extracted to a test helper if similar patterns emerge in other tests
- 🔵 **[INFO]** `provider-swift/Tests/ProviderCoreTests/ContinuousBatchingLiveTests.swift:176-200` — Test method doing multiple validation steps that could be clearer as separate assertions
  - _Suggestion:_ Consider breaking the coherence validation and token matching validation into separate, more focused test methods for better failure isolation

_5 finding(s) total, 1 blocking. Verdict: REQUEST_CHANGES._

_🤖 Automated review by Centaur · DAR-186_
